### PR TITLE
add `std tee` and tests

### PR DESCRIPTION
**Pull request by amtoine**
# Description
this PR adds the `std tee` command to the standard library.
this command does

> read from standard input and write to standard output and files

and is quite common on *nix systems :yum: 

# User-Facing Changes
one can do something like logging multiple steps of a pipeline to a file
```nushell
use std tee
```
```nushell
> ls | tee --force foo | first 2 | tee --append foo | where size > 10kb | tee foo --append | get name | tee foo --append
╭───┬─────────────────╮
│ 0 │ CONTRIBUTING.md │
╰───┴─────────────────╯
```
```nushell
> open foo | lines | each { from nuon }
╭───┬───────────────────────────────────────────────────────────────────────╮
│ 0 │ ╭────┬─────────────────────┬──────┬───────────┬─────────────────────╮ │
│   │ │  # │        name         │ type │   size    │      modified       │ │
│   │ ├────┼─────────────────────┼──────┼───────────┼─────────────────────┤ │
│   │ │  0 │ CODE_OF_CONDUCT.md  │ file │   3.4 KiB │ 2023-04-20 19:41:30 │ │
│   │ │  1 │ CONTRIBUTING.md     │ file │  18.3 KiB │ 2023-08-27 18:45:24 │ │
│   │ │  2 │ Cargo.lock          │ file │ 140.6 KiB │ 2023-08-27 18:45:24 │ │
│   │ │  3 │ Cargo.toml          │ file │   5.6 KiB │ 2023-08-27 18:45:24 │ │
│   │ │  4 │ Cross.toml          │ file │     666 B │ 2023-08-27 18:45:24 │ │
│   │ │  5 │ LICENSE             │ file │   1.1 KiB │ 2023-06-18 11:12:56 │ │
│   │ │  6 │ README.md           │ file │  11.8 KiB │ 2023-07-28 16:40:50 │ │
│   │ │  7 │ assets              │ dir  │   4.0 KiB │ 2023-07-28 16:40:50 │ │
│   │ │  8 │ benches             │ dir  │   4.0 KiB │ 2023-08-27 18:45:24 │ │
│   │ │  9 │ codecov.yml         │ file │     293 B │ 2023-03-11 19:26:57 │ │
│   │ │ 10 │ crates              │ dir  │   4.0 KiB │ 2023-07-28 16:40:50 │ │
│   │ │ 11 │ docker              │ dir  │   4.0 KiB │ 2023-08-27 18:45:24 │ │
│   │ │ 12 │ foo                 │ file │   1.6 KiB │ 2023-08-27 19:32:26 │ │
│   │ │ 13 │ rust-toolchain.toml │ file │   1.1 KiB │ 2023-08-27 18:45:24 │ │
│   │ │ 14 │ scripts             │ dir  │   4.0 KiB │ 2023-08-27 18:45:24 │ │
│   │ │ 15 │ src                 │ dir  │   4.0 KiB │ 2023-08-27 18:45:26 │ │
│   │ │ 16 │ target              │ dir  │   4.0 KiB │ 2023-08-10 10:45:10 │ │
│   │ │ 17 │ tests               │ dir  │   4.0 KiB │ 2023-08-04 14:28:58 │ │
│   │ │ 18 │ toolkit.nu          │ file │  14.6 KiB │ 2023-08-27 18:45:26 │ │
│   │ │ 19 │ wix                 │ dir  │   4.0 KiB │ 2023-07-28 16:40:50 │ │
│   │ ╰────┴─────────────────────┴──────┴───────────┴─────────────────────╯ │
│ 1 │ ╭───┬────────────────────┬──────┬──────────┬─────────────────────╮    │
│   │ │ # │        name        │ type │   size   │      modified       │    │
│   │ ├───┼────────────────────┼──────┼──────────┼─────────────────────┤    │
│   │ │ 0 │ CODE_OF_CONDUCT.md │ file │  3.4 KiB │ 2023-04-20 19:41:30 │    │
│   │ │ 1 │ CONTRIBUTING.md    │ file │ 18.3 KiB │ 2023-08-27 18:45:24 │    │
│   │ ╰───┴────────────────────┴──────┴──────────┴─────────────────────╯    │
│ 2 │ ╭───┬─────────────────┬──────┬──────────┬─────────────────────╮       │
│   │ │ # │      name       │ type │   size   │      modified       │       │
│   │ ├───┼─────────────────┼──────┼──────────┼─────────────────────┤       │
│   │ │ 0 │ CONTRIBUTING.md │ file │ 18.3 KiB │ 2023-08-27 18:45:24 │       │
│   │ ╰───┴─────────────────┴──────┴──────────┴─────────────────────╯       │
│ 3 │ ╭───┬─────────────────╮                                               │
│   │ │ 0 │ CONTRIBUTING.md │                                               │
│   │ ╰───┴─────────────────╯                                               │
╰───┴───────────────────────────────────────────────────────────────────────╯
```

when the `foo` file already exists, the following error is thrown
```nushell
Error:   × std::tee::can_not_dump_to_file
   ╭─[entry #3:1:1]
 1 │ "woopsie" | tee foo
   ·                 ─┬─
   ·                  ╰── file already exists, use `tee --force` to overwrite or `tee --append`
   ╰────
```

# Tests + Formatting
this PR adds the `tee` test to the main `test_std.nu` test module.

# After Submitting

### CODE REVIEW

File: crates/nu-std/std/mod.nu [hunk 1]
```diff
@@ -294,3 +294,35 @@ Startup Time: ($nu.startup-time)
...
     $env.PWD
 }
+
+# read from standard input and write to standard output and files
+export def tee [
+    filename: path  # the file to write the input to
+    --append (-a): bool  # append to the file instead of overwriting
+    --force (-f): bool  # force the overwriting of the file's content
+]: any -> any {
+    let data = $in
```
Comment (sholderbach) [inline crates/nu-std/std/mod.nu:304]: This makes the whole thing not streaming based on how `$in` works.

**Review by sholderbach — COMMENTED** [review]
This sounds like a good addition to satisfy expectations from folks coming from POSIX. I support experimenting and gathering feedback for that with a stdlib implementation.

But I see two potential concerns with such a Nu implementation. This `tee` does not stream and is not transparent in the regular pipeline direction. Meaning you can not use it to live observe a long running operation while writing to a file, also you loose metadata or behavior of a `PipelineData` with some buffering. (e.g. `ls` metadata will be lost)

**Comment by Zoybean** [comment]
As I outlined in #10422, I would love for `nu` to have a `tee` command that doesn't just write to files, but can duplicate its input stream to arbitrary pipelines. But, my idea introduces several complications that don't come up with the file-only design, so I accept if this is a good first (or only) step.

File: crates/nu-std/std/mod.nu (hunk 1 above)
Comment (WindSoilder) [reply crates/nu-std/std/mod.nu:337]: What about making it works with coreutils?

Comment (amtoine) [reply crates/nu-std/std/mod.nu:337]: true, there is a [`tee` command in `uutils`](https://github.com/uutils/coreutils/blob/main/src/uu/tee/tee.md) :open_mouth:

**Comment by amtoine** [comment]
i don't have time to work on this right now...
so i'll go ahead and close this, it will still be available in the list of my closed PRs :relieved:

**Pull request closed without merging**